### PR TITLE
Add CoreDNS details to the customize DNS doc

### PR DESCRIPTION
**Pull request by rajansandeep**
Rewrite the document to include more details about CoreDNS, since it's now the default from v1.12

### CODE REVIEW

**Comment by rajansandeep** [comment]
@johnbelamaric @chrisohaver @fturib PTAL

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 1]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
 using the addon manager
 [cluster add-on](http://releases.k8s.io/{{< param "githubbranch" >}}/cluster/addons/README.md).
 
-The running DNS Pod holds 3 containers:
+Following Kubernetes v1.12, CoreDNS is declared the default DNS Server instead of kube-dns in kubeadm. You may want to refer
```
Comment (chrisohaver) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: This happened in 1.11 in kubeadm, right?

Comment (johnbelamaric) [reply content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: "instead of kube-dns. In kubeadm, this change was made in Kubernetes 1.11, but other installers have waited until 1.12. You should refer..."

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 2]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
...
-
-## Inheriting DNS from the node
+The DNS server supports forward lookups (A records), service lookups (SRV records), and reverse IP address lookups (PTR
+records) as per the [DNS specifications] (https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/)
 
 When running a Pod, kubelet prepends the cluster DNS server and searches
 paths to the node's DNS settings.  If the node is able to resolve DNS names
 specific to the larger environment, Pods should also be able to resolve.
-But see [Known issues](/docs/tasks/administer-cluster/dns-debugging-resolution/#known-issues).
+But see [Known issues](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/#known-issues).
```
Comment (johnbelamaric) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: I think this was intentionally a relative URL, wasn't it. It should match what other links do (so should your link above).

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 3]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
...
 
 If you don't want this, or if you want a different DNS config for pods, you can
 use the kubelet's `--resolv-conf` flag.  Set this flag to "" to prevent Pods from
 inheriting DNS. Set it to a valid file path to specify a file other than
 `/etc/resolv.conf` for DNS inheritance.
 
-## Configure stub-domain and upstream DNS servers
+## CoreDNS
+
+CoreDNS is a flexible, extensible, authoritative DNS server which directly integrates with the Kubernetes API.
```
Comment (johnbelamaric) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: I think this is a little too marketing-y for this doc.

I would just say:

CoreDNS is a general-purpose authoritative DNS-server that can serve as cluster DNS, complying with...

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 4]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
...
+It can serve as cluster DNS, complying with the [dns specifications]
+(https://github.com/kubernetes/dns/blob/master/docs/specification.md). 
+
+### CoreDNS ConfigMap options
+
+CoreDNS is a DNS server that chains plugins, where each plugin adds a new functionality to CoreDNS. 
+This can be configured by maintaining a [Corefile](https://coredns.io/2017/07/23/corefile-explained/), which is the CoreDNS
+configuration file. You can modify the ConfigMap for the CoreDNS Corefile to change how service discovery works. 
+
+Currently, CoreDNS carries the following default Corefile configuration.
```
Comment (johnbelamaric) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: s/Currently, CoreDNS carries/In Kubernetes, CoreDNS is installed with/

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 5]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
...
+        loadbalance
+    }
+``` 
+The Corefile configuration is based on the following [plugins](https://coredns.io/plugins/) of CoreDNS:
+
+* [errors](https://coredns.io/plugins/errors/): Errors are logged to stdout.
+* [health](https://coredns.io/plugins/health/): Health of CoreDNS is reported to http://localhost:8080/health.
+* [kubernetes](https://coredns.io/plugins/kubernetes/): CoreDNS will reply to DNS queries based on IP of the services and pods of Kubernetes. You can find more details [here](https://coredns.io/plugins/kubernetes/). 
+
+> The `pods insecure` option is provided for backward compatibility with kube-dns. You can use the `pod verified` option, which returns an A record only if there exists a pod in same namespace with matching IP.
```
Comment (johnbelamaric) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: Or `pods disabled` if you don't use pod records.

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 6]
```diff
@@ -28,40 +28,158 @@ DNS is a built-in Kubernetes service launched automatically
...
+* [cache](https://coredns.io/plugins/cache/): This enables a frontend cache.
+* [loop](https://coredns.io/plugins/loop/): Detects simple forwarding loops and halts the CoreDNS process if a loop is found.
+* [reload](https://coredns.io/plugins/reload): Allows automatic reload of a changed Corefile.
+* [loadbalance](https://coredns.io/plugins/loadbalance): This is a round-robin DNS loadbalancer by randomizing the order of A, AAAA, and MX records in the answer.
+
+We can modify the default behavior by modifying this configmap.
+
+### Configuration of Stub-domain and upstream nameserver using CoreDNS
+
+CoreDNS has the ability to configure stubdomains and upstream nameservers using the [proxy plugin](https://coredns.io/plugins/proxy/).
```
Comment (johnbelamaric) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: Probably we should recommend `forward` instead.

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md (hunk 3 above)
Comment (chrisohaver) [reply content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: We already introduce coredns and what it is in another introductory document.  I don't think we need to reintroduce it every time we mention it.  Perhaps we should instead link to that doc from here.

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md (hunk 6 above)
Comment (chrisohaver) [reply content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: We still use `proxy` in the default manifest.

Comment (johnbelamaric) [reply content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: ok

**Comment by zparnold** [comment]
@rajansandeep @chrisohaver @johnbelamaric Is this in a place where it's ready for my final review?

**Comment by rajansandeep** [comment]
@zparnold Yes, this is ready for final review.

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 7]
```diff
@@ -28,28 +28,22 @@ DNS is a built-in Kubernetes service launched automatically
 using the addon manager
 [cluster add-on](http://releases.k8s.io/{{< param "githubbranch" >}}/cluster/addons/README.md).
 
-The running DNS Pod holds 3 containers:
+Following Kubernetes v1.12, CoreDNS is declared the default DNS Server instead of kube-dns. In kubeadm, this change was made
```
Comment (chrisohaver) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: "Following Kubernetes v1.12, CoreDNS is declared the default DNS Server instead of kube-dns. In kubeadm, this change was made in Kubernetes 1.11, but other installers have waited until 1.12. You should refer to the documentation of other installers to verify which is the default DNS server."

Perhaps change this to ...

"As of Kubernetes v1.12, CoreDNS is the recommended DNS Server, replacing kube-dns.  However, kube-dns may still be installed by default with certain Kubernetes installer tools. You can refer to the documentation your installer to know which DNS server is installed by default."

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 8]
```diff
@@ -61,7 +55,131 @@ use the kubelet's `--resolv-conf` flag.  Set this flag to "" to prevent Pods fro
...
+        }
+        prometheus :9153
+        proxy . /etc/resolv.conf
+        cache 30
+        loop
+        reload
+        loadbalance
+    }
+``` 
+The Corefile configuration is based on the following [plugins](https://coredns.io/plugins/) of CoreDNS:
```
Comment (chrisohaver) [inline content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: "is based on" => "includes"

**Comment by zparnold** [comment]
@rajansandeep can you please address @chrisohaver 's concerns, and then I can review

File: content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md [hunk 9]
```diff
@@ -28,28 +28,22 @@ DNS is a built-in Kubernetes service launched automatically
...
 with the flag `--cluster-domain=<default-local-domain>`.
 
-The Kubernetes cluster DNS server is based on the
-[SkyDNS](https://github.com/skynetservices/skydns) library. It supports forward
-lookups (A records), service lookups (SRV records), and reverse IP address
-lookups (PTR records).
-
-## Inheriting DNS from the node
+The DNS server supports forward lookups (A records), service lookups (SRV records), and reverse IP address lookups (PTR
+records) as per the [DNS specifications] (/docs/concepts/services-networking/dns-pod-service/)
```
[the remark this answers is not in the document]
Comment (chrisohaver) [reply content/en/docs/tasks/administer-cluster/dns-custom-nameservers.md]: Maybe rephrase here to ...

>The DNS server supports forward lookups (A records), service lookups (SRV records), and reverse IP address lookups (PTR records), and more. For more information see [DNS for Services and Pods] (/docs/concepts/services-networking/dns-pod-service/).

**Comment by rajansandeep** [comment]
/assign @zparnold 
I've addressed all concerns.
